More elaborate/descriptive error message on the `Use a domain I own` page

spraveenitpro opened the pull request:
#### Changes proposed in this Pull Request

At this moment when a customer attempts to transfer a .nl domain to WordPress.com they get this screen:

![](https://d.pr/i/wHFVHx+)
[Direct Link](https://d.pr/i/wHFVHx+)

`This domain can not be transfered` message is quite generic and should be replaced with a dynamic message that includes their TLD so that the error message is more informative like:
`We don't currently support transfer of .nl domains to WordPress.com.`

![](https://d.pr/i/inWxuO+)
[Direct Link](https://d.pr/i/inWxuO+)

#### Testing instructions

1.  Visit http://wordpress.com/domains/manage/:site
2. Click `Add a domain` button on top right and then click the `Domain I own` drop down.
3. Proceed to search for a .nl based domain and click the `next` button to see two card components, this message would be in the second card component.



Related to #61463

Commit:
Add CodeSee architecture diagram workflow to repository

spraveenitpro committed:
Merge branch 'Automattic:trunk' into trunk

spraveenitpro committed:
Merge branch 'Automattic:trunk' into trunk

spraveenitpro committed:
Merge branch 'Automattic:trunk' into trunk

spraveenitpro committed:
Merge branch 'Automattic:trunk' into trunk

spraveenitpro committed:
Merge branch 'Automattic:trunk' into trunk

spraveenitpro committed:
Merge branch 'Automattic:trunk' into trunk

spraveenitpro committed:
Merge branch 'Automattic:trunk' into trunk

spraveenitpro committed:
Merge branch 'Automattic:trunk' into trunk

spraveenitpro committed:
Changing generic message for failure to transfer .nl domain to a more informative message

spraveenitpro committed:
Removing codesee-arch-diagram.yml file that was committed by mistake

tyxla commented on client/components/domains/use-my-domain/utilities/index.js:
```diff
@@ -12,3 +12,4 @@ export { getTransferSalePriceText } from './get-transfer-sale-price-text';
 export { isFreeTransfer } from './is-free-transfer';
 export { optionInfo } from './option-info';
 export { transferDomainAction } from './transfer-domain-action';
+export { getUnsupportedDomainTransferMessage } from './get-unsupported-domain-transfer-message';
```
Nit: should these be alphabetically sorted?

tyxla reviewed (approved):
Code looks good and this tests well for me! Nice work.

I'd like to defer the final approval to @Automattic/nomado though, because it falls in their domain (pun intended).

leonardost commented on client/components/domains/use-my-domain/utilities/get-option-info.js:
```diff
@@ -35,7 +36,7 @@ export const getDomainTransferrability = ( domainInboundTransferStatusInfo ) =>
 			topText:
 				getTransferRestrictionMessage( domainInboundTransferStatusInfo ) ??
-				optionInfo.transferNotSupported.topText,
+				getUnsupportedDomainTransferMessage( domain ),
```
The `if ( ! result.transferrable ) {` condition does not necessarily mean that a domain has a not supported extension, so I think this isn't the best place to change the message - see my other comments

delputnam replied:
Just to add a bit on to what Leo is proposing, I think you could also remove the condition that checks for `! isDomainTransferrable` and add in a call to `getAvailabilityNotice()` in the `default` case of the switch statement.  Then just set the default transfer content to something like:

```
transferContent = {
    ...optionInfo.transferNotSupported,
    topText: availabilityNotice.message,
};
```

(where `availabilityNotice` is the data you get back from the call to `getAvailabilityNotice()`)

This would also allow us to take advantage of all of the pre-existing availability notice data  so we don't have to re-create them.

spraveenitpro replied:
Thanks @delputnam 

I am calling `getAvailabilityNotice()`  like this:

```
import { getAvailabilityNotice } from 'calypso/lib/domains/registration/availability-messages';
const availabilityNotice = getAvailabilityNotice( domain, availability );
```

And finally setting the default condition like this:

```
default:
			transferContent = {
				...optionInfo.transferNotSupported,
				topText: availabilityNotice.message,
			};
```

leonardost reviewed (changes requested):
Hi @spraveenitpro, thanks for working on this! There's just one issue I think we need to take care of 🙂 

There's actually already a message for unsupported TLD transfers [here](https://github.com/spraveenitpro/wp-calypso/blob/764be2782756d53fe0e7659b9893422bfca6c1d2/client/components/domains/use-my-domain/utilities/get-option-info.js#L126-L140) - it seems the problem is that that message is being overwritten by `domainTransferContent` some lines below [here](https://github.com/spraveenitpro/wp-calypso/blob/764be2782756d53fe0e7659b9893422bfca6c1d2/client/components/domains/use-my-domain/utilities/get-option-info.js#L145-L147).

So while changing the message in the `getDomainTransferrability` function works for this case, I'm not sure it's the correct place to do so. That's because users might try to transfer other domains with different transfer errors and end up receiving an incorrect "TLD unsupported" error message.

My suggestion is to update the default logic in the switch statement [here](https://github.com/spraveenitpro/wp-calypso/blob/764be2782756d53fe0e7659b9893422bfca6c1d2/client/components/domains/use-my-domain/utilities/get-option-info.js#L141-L143) and merge it with the `if` statement right below it. Then it'll become something like:

```
default:
	if ( ! isDomainTransferrable ) {
		transferContent = domainTransferContent;
	} else {
		transferContent = optionInfo.transferNotSupported;
	}
```

or even

```
default:
	transferContent = isDomainTransferrable ? optionInfo.transferNotSupported : domainTransferContent;
```

And then removing the `if` condition [below](https://github.com/spraveenitpro/wp-calypso/blob/764be2782756d53fe0e7659b9893422bfca6c1d2/client/components/domains/use-my-domain/utilities/get-option-info.js#L145-L147).

That way, we don't need to define a new message for this case 🙂

spraveenitpro committed:
Refactor fix for better error message based on team feedback

spraveenitpro committed:
Refactor the fix and clean up the unused variables

leonardost commented on client/components/domains/use-my-domain/utilities/get-option-info.js:
```diff
@@ -138,11 +135,10 @@ export function getOptionInfo( {
-
-	if ( ! isDomainTransferrable ) {
-		transferContent = domainTransferContent;
+			transferContent = {
```
One small suggestion here is to define `availabilityNotice` closer to where it's used so the code is easier to understand 🙂 
```suggestion
			const availabilityNotice = getAvailabilityNotice( domain, availability );
			transferContent = {
```

leonardost reviewed (approved):
@spraveenitpro thank you for the updates, this is looking great! The correct message is shown when trying to transfer unsupported domains to us and the code LGTM 👍 left just a small suggestion.

@delputnam what do you think of the PR?

spraveenitpro committed:
Update client/components/domains/use-my-domain/utilities/get-option-info.js

Define availabilityNotice closer to where it's used so the code is easier to understand

Co-authored-by: leonardost <leonardost@users.noreply.github.com>

spraveenitpro commented:
> 👍 left just a small suggestion.

Thanks @leonardost, I have applied your suggestion. ESlint is now showing this:

![](https://d.pr/i/ybw2Df+)
[Direct Link](https://d.pr/i/ybw2Df+)

tyxla commented:
> > 👍 left just a small suggestion.
> 
> Thanks @leonardost, I have applied your suggestion. ESlint is now showing this:
> 
> ![](https://camo.githubusercontent.com/0281e5d88e4fa614f5a606e8dfe4e4435f9acbbd97cb1f0f3e4be61ec7958485/68747470733a2f2f642e70722f692f7962773244662b) [Direct Link](https://d.pr/i/ybw2Df+)

@spraveenitpro this occurs because we have multiple lines of code within a single `case` block without wrapping curly braces. Wrapping both lines in a curly brace group should fix it and is generally considered a good practice. You can find out more about it here: https://eslint.org/docs/rules/no-case-declarations

spraveenitpro committed:
Refactor the default switch block to wrap braces to comply with eslint suggestions

spraveenitpro commented:
> You can find out more about it here: https://eslint.org/docs/rules/no-case-declarations

Thanks for this wonderful explanation, I have just pushed as per the eslint suggestion.

delputnam commented on client/components/domains/use-my-domain/utilities/get-option-info.js:
```diff
@@ -137,12 +132,13 @@ export function getOptionInfo( {
-	if ( ! isDomainTransferrable ) {
-		transferContent = domainTransferContent;
+		default: {
+			const availabilityNotice = getAvailabilityNotice( domain, availability );
```
This should be just the `status` field of the availability object.
```suggestion
			const availabilityNotice = getAvailabilityNotice( domain, availability.status );
```

delputnam reviewed (commented):
Just one small change and then I think this one is ok to merge.

You can check the availability status that is returned from the availability endpoint and then trying various values from the `domainAvailability` constant that might be returned.

spraveenitpro committed:
Refactoring to check the availability status that is returned from the availability endpoint

delputnam reviewed (approved):
This LGTM.  @leonardost do you want to have one more look to make sure I'm not missing anything?

leonardost reviewed (approved):
Tested and works as expected, LGTM! Nice job @spraveenitpro 🚀

tyxla reviewed (commented):
Thanks, @spraveenitpro! I'm going to help ship this.
